### PR TITLE
Initial pull request for Resilient join

### PR DESCRIPTION
**Pull request by foyzur**
Resilient join is the next generation Greenplum join operator. It has the following advantages over existing join operators:

1. Support for multiple join algorithms; currently supports hash join and nested loop join and has provision for implementation of other algorithms in the future
2. Better memory management and memory quota compliance even if there is data skew
3. Robust handling of inaccurate stats
4. Spilling as few tuples as possible, maximizing memory utilization
5. Dynamic hash functions to handle skew better
6. Runtime stat derivation and selection of best applicable join algorithm using heuristics and runtime stats
7. Dynamic switching of join sides, if permitted (role reversal)
8. Join as many spilled partitions as can be fit in the memory (partition tuning)
9. Support for different join algorithms for different sets of spilled partitions

This is the very initial commit of Resilient Join. While dedicated tests for this feature is virtually non-existent at this moment, the code does pass all the 
installcheck-good. Given that join is ubiquitous and heavily exercised during installcheck-good, this gives quite a bit of test coverage.

The intent of this pull request is not to push, until all of us agree to push. All the new code is guarded by gucs that are disabled by default. The main intent, however, is to get early feedback on the design as well as the implementation.

### CODE REVIEW

**Comment by hlinnaka** [comment]
This seems to include a whole new implementation of hashing, using murmurhash. That may well be a good idea, but seems completely unrelated to doing "resilient joins". If it's a better hash algorithm, we probably should use it in more places, not just resilient joins. And conversely, having it in this patch makes it harder to review and evaluate the impact of resilient joins, as opposed to changing hash algortihm. Can you split that into a separate pull request, please?

**Comment by hlinnaka** [comment]
This needs more high-level descriptions and comments. At a quick glance, I couldn't figure out when the switching occurs, which piece of code decides on when to switch, or how switching between different kinds of joins is possible without emitting duplicate output tuples or missing some.

**Comment by foyzur** [comment]
@hlinnaka the resilient join is not about hash function. The hash function is just a supporting API to let the hash join part of the operator work. murmurhash is orthogonal to the implementation of the operator. It's just one family of hash functions that we decided to use. It is trivial to swap out murmurhash and use some other hashing function. 

To make the resilient join work, we need an optimizer that can generate a resilient join plan. The translator layer has been modified to generate a resilient join plan only if the guc enable_resilient_join is turned on. Note, a plan with resilient join looks different than regular hash join as the regular hash join has a HashJoin operator and a child Hash operator, but ResilientJoin is a self contained operator and can assume hash or NLJ during runtime.

Once a plan is generated by the optimizer with a ResilientJoin node, the executor starts executing just like any other operator. This is a recursive join operator where the root operator is set to hash join by default. Therefore, resilient join will pull in all the inner tuples and build a hash table. During this process, it may spill a bunch of buckets and create a bunch of spilled "partitions" (aka files). Once build phase is done, it will start the probe phase and join the probe tuples ONLY for the build partitions that are memory resident and write the rest of the probe tuples into spilled partitions. At the end of this pass we have completed partial join for all the in-memory build tuples (and their corresponding probe tuples). 

The next phase is to read the spill files of both sides, do partition tuning by merging multiple build and probe partitions into one partition (partition tuning) as much as the available memory quota allows. It then finds the smaller of the two "tuned" (combined) partitions and switches the build and probe side (if probe happens to be smaller than the build side, for optimizer wrong stat, or for recursive sub-join, or for whatever other reason, it will swap the roles and probe may become the build side for next recursive sub-join) if the join type allows (e.g., no outer join). A recursive sub-join is then invoked, still using a "hash" type join algorithm.

At the end of a recursive sub-join the resilient join operator may decide that the sub-partitioning process failed to gain much (i.e., too much skew), and may apply other heuristics to decide that hash join is not the most effective way to proceed. It can then change the join algorithm (e.g., using NLJ) for the next recursive join. Once the NLJ (or other recursive join) returns, the parent joiner may switch back to hash join if it finds the next set of tuned (combined) spilled partitions suitable for hash join.

There are couple of gucs in this PR:

1. enable_resilient_join: if set to true, and the optimizer=on then optimizer will generate a plan with resilient join, instead of traditional hash join. Note, the resilient join does not have a "hash" child like a hash join operator. For a resilient join plan generated by optimizer our installcheck-good tests would fail because of plan diff.
2. resilientjoin_spill_strategy: A guc that decides which of the hash buckets to spill if we cannot fit all the build tuples in-memory. SPILL_LARGEST_BUCKETS
would spill the largest buckets (in terms of the size of the hashed tuples). SPILL_LARGEST_MEMORY_CONTEXT is an optimization where instead of finding largest
buckets, it would maintain a fix number of memory contexts where each context hosts multiple buckets and reset the largest memory context, spilling all the 
buckets contained in that context. This is for performance optimization and to minimize creating a heap to find the largest from a very large number of buckets.
3. enable_resilient_join_partition_tuning: Whether to try to combine multiple spill files into one
4. force_new_join: A guc that preserves the existing plan shapes and can test resilient join even without optimizer. If this guc is set, the executor
will automatically convert a hash join plan into a resilient join execution plan. This is done during ExecInitNode, where instead of creating a HashJoin
execution node, we create a ResilientJoin execution node. A Hash node is transparently bypassed using a "HashDummy" node that is just a passthru. This way
we can verify that the ResilientJoin is functionally complete by executing the entire installcheck-good and exercising ResilientJoin every time we have a
HashJoin. This preserves the plan shape as well as to the EXPLAIN system it's just a HashJoin and a Hash child combo.

**Comment by hlinnaka** [comment]
On 12/01/16 23:56, foyzur wrote:
> @hlinnaka the resilient join is not about hash function. The hash
> function is just a supporting API to let the hash join part of the
> operator work. murmurhash is orthogonal to the implementation of the
> operator. It's just one family of hash functions that we decided to
> use. It is trivial to swap out murmurhash and use some other hashing
> function.

Right, so please use the normal hash operators.

> To make the resilient join work, we need an optimizer that can
> generate a resilient join plan. The translator layer has been
> modified to generate a resilient join plan only if the guc
> enable_resilient_join is turned on.

It would be really useful for testing purposes if this wasn't ORCA-only. 
If I understood correctly, a resilient join is basically a hash join, 
with some smarts on how to deal with spills. So I'd suggest always using 
resilient joins instead of hash joins, when enable_resilient_join=on, 
whether or not you use ORCA.

> [explanation of how the resilient join works at execution]

Please put all this in comments.

> At the end of a recursive sub-join the resilient join operator may
> decide that the sub-partitioning process failed to gain much (i.e.,
> too much skew), and may apply other heuristics to decide that hash
> join is not the most effective way to proceed. It can then change the
> join algorithm (e.g., using NLJ) for the next recursive join. Once
> the NLJ (or other recursive join) returns, the parent joiner may
> switch back to hash join if it finds the next set of tuned (combined)
> spilled partitions suitable for hash join.

Hmm. If I understood correctly, this all happens after you have already 
scanned both inner and outer sides once, and constructed the spill files 
for anything that couldn't be joined with in the first in-memory batch. 
How do you perform a nested loop join on the spill files? Do you 
essentially perform a seqscan of the inner spill file for each row in 
the outer spill file? How can that ever be better than doing a probe 
using the hash value?

There's a constant JoinAlgorithm_SortMerge in the patch, but AFAICS it's 
not used for anything. I presume that's a leftover, and the patch 
doesn't do merge joins, only hash and nested loop joins?

Performance-wise, how does this compare with the "skew bucket" mechanism 
in PostgreSQL 8.4? If I understood correctly, that's targeting roughly 
the same use case, where you have highly skewed input data.

- Heikki

File: src/backend/executor/execResilientJoin.c [hunk 1]
```diff
@@ -0,0 +1,1108 @@
...
+ *
+ * ******************** PRIVATE (STATIC) METHODS ***********************************
+ */
+
+/*
+ * Central state transition gateway to decide which
+ * next function to call
+ */
+static TupleTableSlot*
+DelegateToNextStage(ResilientJoiner *resilientJoiner)
```
Comment (foyzur) [inline src/backend/executor/execResilientJoin.c]: The central controller that controls the transitions between different join stages.

File: src/backend/executor/execResilientJoin.c [hunk 2]
```diff
@@ -0,0 +1,1108 @@
...
+{
+	ResilientJoiner *joiner = (ResilientJoiner*)hashJoiner;
+
+	if (NULL == joiner->parentJoiner)
+	{
+		return JoinAlgorithm_HashJoin;
+	}
+
+	/* TODO: Enable NLJ switching */
+	if (false /* Disable NLJ switching */ &&
```
Comment (foyzur) [inline src/backend/executor/execResilientJoin.c]: The current version has NLJ switching disabled as we are yet to implement a suitable heuristic to decide when to give up trying to join using HashJoin. But the NLJ switching works fine as NLJ is just a basic join that uses the same ExecuteSimpleJoin() as the corresponding HashJoin.

**Comment by hlinnaka** [comment]
This doesn't seem to be going anywhere in its current form, so closing.

**Pull request closed without merging**